enhancements/network/ovn-on-hypershift.md

```diff
@@ -0,0 +1,275 @@
+
+Thus, the OpenShift OVN deployment model must be updated to support split control planes. Likewise, the Network Operator must be able to manage components on both the control plane as well as the hosted cluster.
+
+## Motivation
+
+We will elide the full motivation for hosted control planes (i.e. the Hypershift model). However, hosting the OVN control plane off-cluster has the same advantage as hosting the Kubernetes control plane:
+- potentially higher reliability
+- no special master nodes.
+- no chicken-egg problem while bootstrapping (i.e. cannot use most cluster features until networking comes up)
+- More available features (e.g. volumes, stateful sets)
+- etc.
+
+The most significant motivation is the availability of StatefulSets for the OVN databases. In OpenShift,
+we use the ovsdb raft replication mode to provide high availability on master nodes. Managing this
+manually is error prone, especially in the case of lost masters. However, self-hosted OVN cannot use
+StatefulSets, since they require a functioning network. OVN on Hypershift does not suffer this
+chicken-and-egg problem.
+
+### Goals
+
+Hypershift clusters are easily deployed using OVN, with the entire control plane residing off-cluster. OVN is the default networking provider for Hypershift.
+
+### Non-Goals
+
+- Removing the OVN south-bound DB (sbdb) as the primary integration point (for nodes)
+- Full network connectivity between the management and hosted clusters
+## Proposal
+
+The gist of the design is:
+
+1. All OVN control-plane components (ovnkube-master, NB/SB databases, ovn-northd), as well as the CNO, reside on the management cluster
+2. The CNO manages components in both clusters
+3. Communication between the nodes and the OVN control plane is via the southbound OVN database, exposed from the management cluster via a Route
+4. No changes to OVN are required
+
+
+### User Stories
+
+I, a Hypershift user, find running Kubernetes control planes terribly boorish. I would like all the fancy features
+available to me through ovn-kubernetes (EgressIP, EgressFirewall, ICNI, et cetera) without having to
+run that pesky control plane. My cluster networking Just Works without causing me any problems.
+
+### API Extensions
+
+n/a
+
+## Design Details
+
+### Cluster Network Operator (CNO) design
+
+The majority of this proposal's implementation effort will be in CNO. CNO is a complicated operator, with 10 controllers operating in varying levels of the stack.
+The full list of controllers within the CNO can be found [here](https://github.com/openshift/cluster-network-operator/blob/master/docs/operands.md).
+
+Most controllers within the CNO will not need to be aware of the management cluster, and will thus remain unchanged. All of their operands are entirely
+on the hosted cluster and Hypershift will be entirely transparent.
+
+Work needed to CNO includes:
+- Make specific controllers aware of multiple clusters
+- Refactor controllers off of controller-framework
```

Comment:
I am curious why this is a requirement.

Reply:
Mostly because of the difficulty of wedging muli-cluster support in to controller-framework. Also, controller-framework has never been a good fit for CNO - we would like the ability to do server-side label scoped watches, for example.

Reply:
If you mean controller-runtime, it does support both multi-cluster controllers and label-filtered watches. But either way, such a refactoring seems to be a bit orthogonal?

Reply:
There are a few places where it doesn't really fit our needs. Agreed that this is mostly orthogonal at this point.